Feature/get tests elasticsearch source

adrianfusco opened the pull request:
- Adding method get_tests for elasticsearch_jobs
- I've realized that some documents have bad information registered in the test_time field. Not just the time but more information. For now I'll catch this and avoid it when it happens (there's a task open).

JoseSilvaHermida reviewed (commented):
I tested it and it works, but I noticed it shows all the builds with all their tests.

The implementation for Jenkins just returns the tests for the last completed build of each job.
I don't know which one of the two approaches is the preferred one... any suggestions?

cescgina commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -95,16 +96,14 @@ def __query_get_hits(self: object, query: dict, index: str = '*') -> list:
-                    size=10000
-                )]
-            es_connection.transport.close()
+            LOG.info("Using the following query: {}"
```
I think this should be moved to debug

adrianfusco replied:
Done.

adrianfusco commented:
> I would say that showing the tests for last completed build is probably the most useful option for most cases, particularly if the user is more interested on which tests are run. However, it maybe could be useful to see the result of a test for all builds, so maybe we want to support that too. We can raise the topic in the weekly if there is time.

@JoseSilvaHermida @cescgina 

Unfortunately there are documents stored that has the tests and the build number but not the build status.

cescgina commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -355,6 +354,98 @@ def get_deployment(self, **kwargs):
+            try:
+                test_duration = float(hit['_source']['test_time'])*1000
+            except ValueError:
+                LOG.warning("'test_time' field is not well parsed in "
```
This line should probably go to debug too, I did a query for just one job and I got 200 of this warnings. That's fine at debug level, but not sure if we should show that in a normal run

adrianfusco replied:
Done.

cescgina commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -95,16 +96,14 @@ def __query_get_hits(self: object, query: dict, index: str = '*') -> list:
-                    size=10000
-                )]
-            es_connection.transport.close()
+            LOG.debug("Using the following query: {}"
```
Flake8 is complaining about the indentation of the second line. With logging you can use lazy old-style formatting, like this:
```python
LOG.debug("Using the following query: %s",
          str(query).replace("'", '"')) 
```
which pylint also prefers

adrianfusco replied:
I didn't have time to check it. Change done.

bregman-arie commented:
> > I would say that showing the tests for last completed build is probably the most useful option for most cases, particularly if the user is more interested on which tests are run. However, it maybe could be useful to see the result of a test for all builds, so maybe we want to support that too. We can raise the topic in the weekly if there is time.
> 
> @JoseSilvaHermida @cescgina
> 
> Unfortunately there are documents stored that has the tests and the build number but not the build status.

I think we shouldn't support `cibyl --tests` by itself. It should probably require the user to add job name.

adrianfusco commented:
> > > I would say that showing the tests for last completed build is probably the most useful option for most cases, particularly if the user is more interested on which tests are run. However, it maybe could be useful to see the result of a test for all builds, so maybe we want to support that too. We can raise the topic in the weekly if there is time.
> > 
> > 
> > @JoseSilvaHermida @cescgina
> > Unfortunately there are documents stored that has the tests and the build number but not the build status.
> 
> I think we shouldn't support `cibyl --tests` by itself. It should probably require the user to add job name.

The problem isn't the job name. Actually we're testing it using a single job name. The thing is if we want to show all the tests of an specific build, all builds or the last successful build. We need to clarify it to adjust all the sources because there's a discrepancy right now.

bregman-arie commented:
> > > > I would say that showing the tests for last completed build is probably the most useful option for most cases, particularly if the user is more interested on which tests are run. However, it maybe could be useful to see the result of a test for all builds, so maybe we want to support that too. We can raise the topic in the weekly if there is time.
> > > 
> > > 
> > > @JoseSilvaHermida @cescgina
> > > Unfortunately there are documents stored that has the tests and the build number but not the build status.
> > 
> > 
> > I think we shouldn't support `cibyl --tests` by itself. It should probably require the user to add job name.
> 
> The problem isn't the job name. Actually we're testing it using a single job name. The thing is if we want to show all the tests of an specific build, all builds or the last successful build. We need to clarify it to adjust all the sources because there's a discrepancy right now.

So I can say for sure that we don't want to show the tests of all builds.
Let's choose between last successful build or last build and make sure that we align this behavior (and test it) across sources.

At the same time please make sure with Jose that `cibyl --tests` will throw an exception to the user asking it specify specific job(s)

adrianfusco commented:
- Added new `Argument` in the `arguments` list of the `Job` class with the name `--build-number`.
- Added filtering by `--build-number` in `get_tests` and `get_builds`
- Added more tests.

jsanemet commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -355,6 +364,125 @@ def get_deployment(self, **kwargs):
+        if 'build_number' in kwargs:
+            build_numbers = [status.upper()
+                             for status in
+                             kwargs.get('build_number').value]
```
If 'build_number' is not on kwargs, this will crash because you will try to access an attribute on a None object.

adrianfusco replied:
Can you clarify this case and the command?
```
cibyl -vvv -d --sources elasticsearch --jobs **name_test** --tests
```
It doesn't crash and it's implemented in other previous methods.

jsanemet replied:
Sorry, my bad. You already checked that the argument is there on line 406, so what I said is not going to happen.

jsanemet commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -355,6 +364,125 @@ def get_deployment(self, **kwargs):
+                }
+            }
+            for build_number in build_numbers:
+                query_body['query']['bool']['must'][1]['bool']['should'] = {
```
Be careful using index access so freely. If just one of these indexes does not exist, then you will get an AttributeError out of here.

adrianfusco replied:
There will no be problem with this, I replace always the same index that is a dictionary that contains a DSL `match` directive.

jsanemet replied:
Ok, if you know it will not bring trouble you can close this one down.

jsanemet commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -355,6 +364,125 @@ def get_deployment(self, **kwargs):
+                             kwargs.get('build_number').value]
+
+        results = []
+        hits = []
```
As far as I can tell, 'hits' and 'results' are copies of one another. I think you can get rid of one of them.

adrianfusco replied:
You're right. Before I needed to access to some index of `results` now it's not necessary. I deleted it.

jsanemet commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -355,6 +364,125 @@ def get_deployment(self, **kwargs):
+        return AttributeDictValue(
+            "jobs",
+            attr_type=Job,
+            value=jobs_found
```
'jobs_found' is already an 'AttributeDictValue'. This output will not be usable.

jsanemet commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -355,6 +364,125 @@ def get_deployment(self, **kwargs):
+                        hits.append(result)
+
+        if not results:
+            return jobs_found
```
Why bother with this check? The next loop will pass through if 'len(hits) == 0'.

adrianfusco replied:
Done.

jsanemet reviewed (changes requested):
I have added a couple comments regarding the code itself. Mostly details, but there a few that should be reviewed.

adrianfusco commented on cibyl/models/ci/job.py:
```diff
@@ -47,7 +47,10 @@ class Job(Model):
                                    func='get_builds', nargs=0,
-                                   description="Last build for job")]
+                                   description="Last build for job"),
+                          Argument(name='--build-number', arg_type=str,
```
Then I understand this situations is similar when we delete `--job-name` and we just used `--jobs`. Is it right?

cescgina replied:
Yes, there was a --build-id argument that was deprecated in favor of --builds. As @jsanemet  said, the --build-number is not really necessary

adrianfusco replied:
Thanks for clarify, I'll delete it and implement in the other way

adrianfusco commented:
- Changed call done from `get_tests` from `get_jobs` to `get_builds`:

1. We'll avoid one call.
2. The `get_builds` method will handle the logic according to the arguments passed: `--builds`, `--builds num_1 num_2 ...`, `--last-build`

cescgina commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -355,6 +365,116 @@ def get_deployment(self, **kwargs):
+                }
+            }
+            builds = job_builds_found[job].builds
+            for build_number, _ in builds.items():
```
Since you just want the dictionary key you can simply iterate with
```python
for build_numb in builds:
```

adrianfusco replied:
You're right. Changed! Thx.

cescgina commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -355,6 +365,116 @@ def get_deployment(self, **kwargs):
+                          )
+                continue
+
+            build_object = Build(
```
Do you need to create this Build object? The call to get_builds should ensure that is already created, shouldn't it? If so, you could skip lines 462-466 and just add the test

adrianfusco replied:
Since I changed the call from `get_jobs` to `get_builds` and we don't have calls to Elasticsearch to get the builds from this method, no, we don't need this build object. Thx. Change done.

JoseSilvaHermida commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -355,6 +365,116 @@ def get_deployment(self, **kwargs):
+                build_id=build_number
+            )
+
+            job_builds_found[job_name].add_build(build_object)
```
I tested it
`$ cibyl --jobs DFG-network-networking-ovn-17.0_director-rhel-virthost-3cont_2comp_1ipa-ipv4-vlan-tls --tests --last-build
INFO     cibyl.sources.elasticsearch.client Connection established successfully with elasticsearch instance: http://seal52.qa.lab.tlv.redhat.com:9200
INFO     cibyl.plugins        Loading plugin: openstack
INFO     cibyl.orchestrator   Took 8.00s to query system osp_jenkins using source elasticsearch of type elasticsearch using method get_tests
Environment: osp_phases
  System: osp_jenkins
      Job: DFG-network-networking-ovn-17.0_director-rhel-virthost-3cont_2comp_1ipa-ipv4-vlan-tls
        Build: 13
          Status: FAILURE
        Build: 13
          Test: test_create_subnet_when_quotas_is_full[id-0b7f99e3-9f77-45ce-9a89-b39a184de618,negative]
            Result: SUCCESS
`
This output shows Build 13 twice.
I think this line triggers that, can you review it?

adrianfusco replied:
Change done: https://github.com/rhos-infra/cibyl/pull/189#discussion_r862707186

cescgina commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -136,6 +136,12 @@ def get_builds(self: object, **kwargs: Argument):
                 build_statuses = [status.upper()
                                   for status in
                                   kwargs.get('build_status').value]
+            build_numbers = []
```
This can be deleted I think

adrianfusco replied:
Since `--build-number` has been deleted, yes, we can remove it. I didn't realize to do this change from the moment it wasn't deleted of the `Job` class. Done.

cescgina commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -154,6 +160,10 @@ def get_builds(self: object, **kwargs: Argument):
                         build['_source']['build_result'] not in build_statuses:
                     continue
 
+                if 'build_number' in kwargs and \
```
This is also unnecessary

adrianfusco replied:
Since `--build-number` has been deleted, yes, we can remove it. I didn't realize to do this change from the moment it wasn't deleted of the `Job` class. Done.

cescgina reviewed (changes requested):
Just need to remove a few lines of dead code and should be ready to merge

cescgina reviewed (approved):
Looks great, thanks for the changes.

adrianfusco commented:
Thanks to everyone for the reviews.
